### ISSUE TITLE
[Scopes] populate promises

### ISSUE THREAD

**Issue by clarkbw** [devtools-html/debugger.html#374]
Promises have special properties we want to display in the variables view for inspection.

*(this is one of the cooler bugs imo)*

#### Old Debugger
Here is where the [old debugger](https://dxr.mozilla.org/mozilla-central/source/devtools/client/shared/widgets/VariablesView.jsm#3638-3645) got the properties.

Here's what the code looked like:
```js
    if (aGrip.class == "Promise" && aGrip.promiseState) {
      let { state, value, reason } = aGrip.promiseState;
      props.push("<state>: " + VariablesView.getString(state));
      if (state == "fulfilled") {
        props.push("<value>: " + VariablesView.getString(value, { concise: true }));
      } else if (state == "rejected") {
        props.push("<reason>: " + VariablesView.getString(reason, { concise: true }));
      }
    }
```

#### Sketch

```diff
diff --git a/src/components/Scopes.js b/src/components/Scopes.js
index 97ed485..d6e93bf 100644
--- a/src/components/Scopes.js
+++ b/src/components/Scopes.js
@@ -72,6 +72,16 @@ function getThisVariable(frame, path) {
   };
 }

+// i could be wrong about formatting each variable...
+function getPromiseProperties(variable) {
+  if (variable.class != "Promise" || !variable.promiseState) {
+    return variable;
+  }
+
+  // old debugger code here
+  // <value>, <reason>, <state>
+}
+
 function getScopes(pauseInfo, selectedFrame) {
   if (!pauseInfo || !selectedFrame) {
     return null;
@@ -116,6 +126,8 @@ function getScopes(pauseInfo, selectedFrame) {
       }

       if (vars && vars.length) {
+        vars = vars.map(variable => getPromiseProperties(variable));
+
         vars.sort((a, b) => a.name.localeCompare(b.name));
         scopes.push({ name: title, path: key, contents: vars });
       }
```

#### Demo code
I made this demo with this example. Maybe it was stupid, but I added the promises to todo-view#toggleCompleted for fun :)

```diff
diff --git a/src/test/examples/todomvc/js/views/todo-view.js b/src/test/examples/todomvc/js/views/todo-view.js
index 0b9e93d..adc1c5c 100644
--- a/src/test/examples/todomvc/js/views/todo-view.js
+++ b/src/test/examples/todomvc/js/views/todo-view.js
@@ -67,6 +67,13 @@ var app = app || {};

                // Toggle the `"completed"` state of the model.
                toggleCompleted: function () {
+                 let y = new Promise(() => {});
+                 let z = new Promise((resolve) => resolve(2));
+                 let x = new Promise((r, reject) => reject(3));
+
+                 return new Promise(function(resolve, reject) {
+                         debugger
+                 })
                        this.model.toggle();
                },
```

#### Firefox
<img width="1081" alt="screen shot 2016-11-19 at 7 57 22 am" src="https://cloud.githubusercontent.com/assets/254562/20455612/f5811b2c-ae2d-11e6-947a-e30f15cae39f.png">

#### Chrome
![screen shot 2016-11-19 at 7 58 10 am](https://cloud.githubusercontent.com/assets/254562/20455614/f85d74b2-ae2d-11e6-95e0-abc747169f0f.png)

**Comment by clarkbw** [comment]
@jlongster when you get a chance, can you explain this one in more detail?  

I'm unclear what our current debugger is doing other than providing the default parameter names for the promise argument functions such that it shows `resolve(value)` and `reject(reason)`.

**Comment by jasonLaster** [comment]
Hey @Zacqary I feel like you would enjoy this one :P

**Comment by Zacqary** [comment]
So far I'm starting by trying to replicate the behavior of the old code but I'm running into an issue. I figured, just to get some results on the screen, I'd try this ugly hack involving mutating the preview:
```javascript
function getPromiseProperties(variable) {
  const { class: varClass, promiseState } = variable.contents.value;
  if (varClass === "Promise" && Boolean(promiseState)) {
    variable.contents.value.preview.ownProperties.promiseState =
      promiseState.state;
  }
  return variable;
}
```

The expected result, I'm thinking, is:
```
x: Promise
    promiseState: "rejected"
    __proto__: PromisePrototype
```
but that's not happening. It's still just:
```
x: Promise
    __proto__: PromisePrototype
```

Just to make sure I wasn't missing something, I also tried making sure the changes I was making to the preview exactly matched what another, working, Object preview looked like:
```javascript
function getPromiseProperties(variable) {
  const { class: varClass, promiseState } = variable.contents.value;
  if (varClass === "Promise" && Boolean(promiseState)) {
    variable.contents.value.preview.ownProperties.promiseState = {
      value: promiseState.state,
      configurable: true,
      enumerable: true,
      writable: true
    };
    variable.contents.value.preview.ownPropertiesLength++;
    variable.contents.value.ownPropertyLength++;
  }
  return variable;
}
```

And still nothing.

Of course, as I found out, `variable.contents.value.preview` seems to only be relevant for rendering the non-tiny grip preview:
![screen shot 2016-11-20 at 5 15 15 pm](https://cloud.githubusercontent.com/assets/1445834/20466663/fef740d8-af44-11e6-860e-9c9ba1ba4974.png)

So obviously that's not what I'm looking for, but I still haven't been able to track down how to add anything to the actual list of properties that gets rendered when you expand one of the grips.

Side note: it seems really weird to me that there are two sources of truth for what appears in the preview, and what appears when you expand the grip.

**Comment by jasonLaster** [comment]
Hmm. Really interesting. I haven't looked at this either. I'd be happy to
check it out tomorrow morning though

**Comment by jasonLaster** [comment]
Thanks @Zacqary, I just looked into it for a bit and I think we might want to format Promises with the Reps module, which you've already looked at before :)

@janodvarko is there a ticket for using Reps to format promises? It looks like `GripMap` and `GripArray` might be similar, but I couldn't find where they're used.

**Comment by nchevobbe** [comment]
@jasonLaster Hi Jason. The Promise Rep (PromiseRep from reps/promise.js ) already landed some times ago (see https://bugzilla.mozilla.org/show_bug.cgi?id=1310608 ). It looks like you don't have it in debugger.html yet though, merging from mozilla-central should be good.

**Comment by Zacqary** [comment]
Had a private conversation with @jasonlaster about some things that are bugging me about the data model that I was playing with in `getScopes` and we figured I should copy it here:

I think the most frustrating thing was the fact that there's this preview prop which is solely for generating the properties displayed in the non-tiny preview (which is for, I assume, when we display a rep in the console as opposed to the Scopes dropdown?) And that one, by the way, has two different ownPropertiesLength and ownPropertyLength props, which could both simply be computed by getting Object.keys(ownProperties).length or equivalent. It seems to me like there's a lot of precomputed data which would be better served by selector functions. Especially given that there's no clear link between the ownProperties in the preview and the properties that show when you open the dropdown. If it's memoizing all that for performance reasons, then it ought to be doing that using a memoized function that points to the original source of truth, so that it's easier to reason about the code.

Maybe the previews are being calculated server-side, e.g. if there are 100,000 variables in a scope. I can see why that might be the case, and I admit that I'm still coming from a place of ignorance because I still don't actually know where the dropdown props are coming from. That's my main concern. My instinct is that the preview and the dropdown should be different views of the same data, but if that's not possible then I'd still like to see a clearer connection between what's going on in getScopes and the props that appear when you expand a scope's dropdown.

In short, the old code patterns that were lifted wholesale from the old debugger are confusing to me as a seasoned Redux dev, and I think it'd benefit contributors with a Redux/Elm-architecture background if they were rethought with a fresh perspective.

**Comment by nchevobbe** [comment]
Hi @Zacqary !

> Maybe the previews are being calculated server-side

They are. The difference you see between `ownPropertiesLength` and `Object.keys(ownProperties)` is indeed for performance reasons.
I think the preview holds the first 10 items/properties of the object, and `ownPropertiesLength` has the actual number of properties/items on the object, so we can output the 10 first items and a ", N more..." to indicate there's more to see.

Like Jason said, using Reps you shouldn't need to have to worry about how to display a grip (and I think using Reps is nice because we have the same object representation all across the devtools). And if you feel that an object doesn't have the representation it should have, please file a bug and/or ping me 😃.

> My instinct is that the preview and the dropdown should be different views of the same data, but if that's not possible then I'd still like to see a clearer connection between what's going on in getScopes and the props that appear when you expand a scope's dropdown.

I don't know for the debugger "scope" panel, but on other devtools panels (like the DOM one), when the user "expand" an object, it calls the server to get the grip for each properties/items, and then create for each of them a new Rep with the retrieved grip (and so on if the properties/items are themselves expandables). You can check the TreeView component (http://searchfox.org/mozilla-central/source/devtools/client/shared/components/tree/tree-view.js), which use http://searchfox.org/mozilla-central/source/devtools/client/dom/content/grip-provider.js as the "provider" (see how properties are fetched in action:  http://searchfox.org/mozilla-central/source/devtools/client/dom/content/actions/grips.js#45).

That said, I don't know at all if such logic could work on the debugger's scope panels, or even if it is wanted, I just wanted to tell you how similar thing was done in the devtools 😉.

**Comment by jasonLaster** [comment]
ping @Zacqary. Mind if I take this bug back so that I can pull in the latest greatest Reps?

**Comment by Zacqary** [comment]
Sure, go right ahead

**Comment by jasonLaster** [comment]
The rep looks better. Now we just need the fields

![screen shot 2017-01-10 at 4 39 44 pm](https://cloud.githubusercontent.com/assets/254562/21825877/77cc0c1a-d753-11e6-9fc9-7b3892d3520c.png)

**Comment by jasonLaster** [comment]
Here's how chrome and firefox handle it:

|chrome|firefox|
|--|--|
|![][chrome]|![][firefox]|

[chrome]:https://cloud.githubusercontent.com/assets/254562/22755253/631b4866-ee10-11e6-9a67-8fb27ca2df07.png
[firefox]:https://cloud.githubusercontent.com/assets/254562/22755254/631bd830-ee10-11e6-8e26-2a3b1ce9c75f.png

**Comment by jasonLaster** [comment]
@janodvarko What's your timeline?

**Comment by janodvarko** [comment]
I'll try to find some time this week.
Honza

**Comment by jasonLaster** [comment]
took another look at this after looking at object-inspector a bit

```diff
diff --git a/src/utils/object-inspector.js b/src/utils/object-inspector.js
index 8fd893c..8decf19 100644
--- a/src/utils/object-inspector.js
+++ b/src/utils/object-inspector.js
@@ -34,6 +34,18 @@ function isDefault(item) {
   return WINDOW_PROPERTIES.includes(item.name);
 }

+function isPromise(item) {
+  return item.value.class == "Promise";
+}
+
+function promiseItems(item) {
+  return [
+    createNode("state", `${item.path}/state`, item.value.promiseState.state),
+    createNode("reason", `${item.path}/reason`, item.value.promiseState.x),
+    createNode("value", `${item.path}/value`, item.value.promiseState.x)
+  ];
+}
+
 function sortProperties(properties) {
   return properties.sort((a, b) => {
     // Sort numbers in ascending order and sort strings lexicographically
@@ -139,6 +151,10 @@ function getChildren({
     return [];
   }

+  if (isPromise(item)) {
+    return promiseItems(item);
+  }
+
   const actor = obj.value.actor;
   ```
   
   I also wrote some tests that we can use. I found writing tests with real data here really helped:
   
   ```diff
   diff --git a/src/utils/tests/object-inspector.js b/src/utils/tests/object-inspector.js
index 500bbd9..9bb5906 100644
--- a/src/utils/tests/object-inspector.js
+++ b/src/utils/tests/object-inspector.js
 describe("object-inspector", () => {
   describe("makeNodesForProperties", () => {
     it("kitchen sink", () => {
@@ -103,4 +105,70 @@ describe("object-inspector", () => {
       ]);
     });
   });
+
+  describe("getChildren", () => {
+    it("rejected promise", () => {
+      const promise = {
+        "enumerable": true,
+        "configurable": false,
+        "value": {
+          "frozen": false,
+          "ownPropertyLength": 0,
+          "preview": {
+            "kind": "Object",
+            "ownProperties": {},
+            "ownPropertiesLength": 0,
+            "safeGetterValues": {}
+          },
+          "actor": "server2.conn2.child1/pausedobj36",
+          "promiseState": {
+            "state": "rejected",
+            "reason": {
+              "type": "undefined"
+            },
+            "creationTimestamp": 1486584316133.3994,
+            "timeToSettle": 0.001713000237941742
+          },
+          "class": "Promise",
+          "type": "object",
+          "extensible": true,
+          "sealed": false
+        },
+        "writable": true
+      }
+
+      const children = getChildren(x, x, promise);
+      expect(children.map(c => c.path)).to.eql([...])
+    })
+
+    it("unhandled promise", () => {
+      const promise = {
+        "enumerable": true,
+        "configurable": false,
+        "value": {
+          "frozen": false,
+          "ownPropertyLength": 0,
+          "preview": {
+            "kind": "Object",
+            "ownProperties": {},
+            "ownPropertiesLength": 0,
+            "safeGetterValues": {}
+          },
+          "actor": "server2.conn2.child1/pausedobj37",
+          "promiseState": {
+            "state": "pending",
+            "creationTimestamp": 1486584316133.2642
+          },
+          "class": "Promise",
+          "type": "object",
+          "extensible": true,
+          "sealed": false
+        },
+        "writable": true
+      }
+
+      const children = getChildren(x, x, promise);
+      expect(children.map(c => c.path)).to.eql([...])
+    })
+  })
```
